backend/asmgen.ml

```diff
@@ -376,28 +387,90 @@ let compile_unit ~output_prefix ~asm_filename ~keep_asm ~obj_filename ~may_reduc
        if create_asm && not keep_asm then remove_file asm_filename
     )
 
-let end_gen_implementation0 ?toplevel ~ppf_dump make_cmm =
-  emit_begin_assembly ();
+let build_dwarf ~asm_directives:(module Asm_directives : Asm_targets.Asm_directives_intf.S) sourcefile =
+  let unit_name =
+    Compilation_unit.get_persistent_ident (Compilation_unit.get_current_exn ())
+  in
+  Dwarf.create
+    ~sourcefile
+    ~unit_name
+    ~asm_directives:(module Asm_directives)
+    ~get_file_id:(Emitaux.get_file_num ~file_emitter:X86_dsl.D.file)
+
+let build_asm_directives () : (module Asm_targets.Asm_directives_intf.S) = (
+    module Asm_targets.Asm_directives.Make(struct
+
+      let emit_line str = X86_dsl.D.comment str
+
+      let get_file_num file_name =
+        Emitaux.get_file_num ~file_emitter:X86_dsl.D.file file_name
+
+      module D = struct
+        open X86_ast
+
+        include X86_dsl.D
+
+        type data_type =
+          | NONE | DWORD | QWORD
+
+        type nonrec constant = constant
+        let const_int64 num = Const num
+        let const_label str = ConstLabel str
+        let const_add c1 c2 = ConstAdd (c1, c2)
+        let const_sub c1 c2 = ConstSub (c1, c2)
+
+        let label ?data_type str =
+          let typ =
+            Option.map
+              (function
+                | NONE -> X86_ast.NONE
+                | DWORD -> X86_ast.DWORD
+                | QWORD -> X86_ast.QWORD)
+              data_type
+          in
+          label ?typ str
+      end
+    end)
+  )
+
+let end_gen_implementation0 ?toplevel ~ppf_dump ~sourcefile make_cmm =
+  let asm_directives =
+    if !Clflags.debug then
+      Some (build_asm_directives ())
+    else
+      None
+  in
+  emit_begin_assembly (fun () ->
+    Option.iter
+      (fun (module Asm_directives : Asm_targets.Asm_directives_intf.S) ->
+        Asm_targets.Asm_label.initialize ~new_label:Cmm.new_label;
+        Asm_directives.initialize ())
+      asm_directives
+  );
+  let dwarf =
+    Option.map
+      (fun asm_directives -> build_dwarf ~asm_directives sourcefile)
+      asm_directives
+  in
```

Comment:
It would be nice to refactor L437-L454 into a separate function (`asm_directives` isn't used after L454, checking Clflags.debug at the top to avoid Option.iter and Option.map), returning a pair of init_dwarf and dwarf, and maybe shared it with similar code in asmlink. Sorry if this use of nested anonymous functions is intentional, I'm unfamiliar with this pattern.

Reply:
I refactored most of it in a function `emit_begin_assembly_with_dwarf` that calls `emit_begin_assembly` and returns an instance of `dwarf` to be used to generate dwarf information.